pkg/services/auth/auth_token.go

```diff
@@ -404,6 +410,35 @@ func (s *UserAuthTokenService) rotatedAfterParam() int64 {
 	return getTime().Add(-s.Cfg.LoginMaxInactiveLifetime).Unix()
 }
 
+const (
+	concurrentSessionLimit    = 3
+	concurrentSessionLimitKey = "concurrentSessionLimit"
+)
+
+func (s *UserAuthTokenService) isConcurrentSessionLimitEnabled() bool {
+	return s.Cfg.FeatureToggles[concurrentSessionLimitKey]
+}
+
+func (s *UserAuthTokenService) revokeOldUserTokens(ctx context.Context) error {
+	if !s.isConcurrentSessionLimitEnabled() {
+		return nil
+	}
+
+	return s.SQLStore.WithDbSession(ctx, func(dbSession *sqlstore.DBSession) error {
+		result, err := dbSession.Exec("DELETE FROM user_auth_token WHERE id NOT IN (SELECT id FROM user_auth_token WHERE seen_at <> 0 ORDER BY seen_at DESC LIMIT ?) AND seen_at <> 0;", concurrentSessionLimit-1)
```

Comment:
Should filter by user id here. Couldn't this query be written with only delete clause/without `in(...)`, maybe not possible with order by then?

Reply:
Completely agree, I already figured out the filter by user id missing but I was waiting for feedback to do it. Anyway, thanks for pointing it, it helped me to do not forget it.

Regarding the query, I cannot see a simpler way to do it. Probably, the subquery might be replaced by a JOIN clause but not really sure if in a simpler / clearer way. Feel free to write down an example if you have something in mind, I'd really appreciate it.

Thanks!

Reply:
Related with the query ☝️ I already was exploring the approach using something like `DELETE ... LIMIT ... OFFSET` which might be enough for our case and also [seem to be supported by SQLite](https://www.sqlite.org/compile.html#enable_update_delete_limit) (also by PosgreSQL & MySQL I guess), but it [didn't seem to be supported by the Go client](https://github.com/mattn/go-sqlite3/issues/787).